File: worker/awards/labels.go
// Package awards is the roulette PRODUCER side of the Provably-Fair draw engine (Story 6.3,
// FR-25 / AD-14). It holds the deterministic, integer-only PRNG primitives the ceremony draws
// from: domain-separated labels, the HMAC-SHA256 counter-mode block stream, and the unbiased
// uniform_int.
//
// THE SEAM (ARCHITECTURE-SPINE.md:73-76). There is NO dependency edge between worker/* and
// app/+lib/. This package is the producer; lib/roulette is the browser VERIFIER. They conform to
// `roulette/vectors/` INDEPENDENTLY — never to each other. Neither is the reference
// implementation and NEITHER MAY BE CORRECTED BY READING THE OTHER'S SOURCE: when they disagree
// the vector decides, and when the vector is silent, add a vector first.
//
// This package is a LEAF: it imports nothing from worker/ingest, worker/store or worker/db. The
// producer wiring that reads the AD-19 snapshot and writes results is Story 6.4+; 6.3 ships the
// primitive and its tests only.
//
// INTEGER-ONLY (SPEC Constraint 7). No float64, no math/rand, no time, no locale-aware
// formatting anywhere in the draw path. The whole point is that a browser can reproduce every
// byte; anything that varies by platform, clock or locale breaks that.
package awards

import (
	"fmt"
	"strconv"
)

// The domain separators. Each label is an INDEPENDENT stream that starts at counter i = 0, so a
// Stage-1 spin can never consume bytes that a later pity draw expects (and vice versa).
//
// AD-14 / SOLUTION-DESIGN §9.1 fix these strings exactly:
//
//	"inclusivcup/v1/stage1/spin/<S>"   S = 1,2,3,…  1-based decimal, NO padding
//	"inclusivcup/v1/pity"
//
// The "v1" is the algorithm version and is part of the domain separation: bumping it is a
// deliberate, ceremony-invalidating act, not a refactor.
const (
	// labelPrefix is the versioned namespace shared by every stream.
	labelPrefix = "inclusivcup/v1"

	// PityLabel is the single pity stream. Story 6.3 defines the LABEL only; the pity
	// ALGORITHM (FR-26's pity roulette) is Story 6.7's.
	PityLabel = labelPrefix + "/pity"
)

// Stage1Label returns the domain separator for Stage-1 spin S.
//
// spin is 1-BASED (spin 1 is the first award drawn) and is formatted as plain decimal with no
// zero padding: spin 1 -> ".../spin/1", spin 12 -> ".../spin/12". strconv.Itoa is used rather
// than fmt.Sprintf precisely because it cannot be given a width/flag by a later edit and is not
// locale-aware — a padded ".../spin/01" is a DIFFERENT stream that would still look correct.
//
// A non-positive spin is a programmer error (there is no spin 0 and no negative spin), so it
// returns an error rather than producing a label that would silently key a valid-looking stream.
func Stage1Label(spin int) (string, error) {
	if spin < 1 {
		return "", fmt.Errorf("awards: stage1 spin must be >= 1 (1-based), got %d", spin)
	}
	return labelPrefix + "/stage1/spin/" + strconv.Itoa(spin), nil
}

File: lib/roulette/labels.ts
/**
 * Roulette domain-separation labels — the browser VERIFIER's mirror of `worker/awards/labels.go`
 * (Story 6.3, FR-25 / AD-14).
 *
 * ⛔ THIS MODULE DELIBERATELY HAS NO `import 'server-only'`. Every other module under `lib/**` is
 * server-only; `lib/roulette` is the first one that is not, because it is the FIRST
 * client-reachable module in `lib/` — 6.9's "Verificar la ceremonia" button pulls it into the
 * browser bundle, and `server-only` throws there. Adding the import out of habit would not fail
 * here; it would fail in someone else's story. `prng.test.ts` pins its absence by scanning this
 * file's source text (the Vitest alias stubs `server-only`, so an import-based test cannot tell
 * the difference).
 *
 * THE SEAM (ARCHITECTURE-SPINE.md:73-76). This module conforms to `roulette/vectors/`, never to
 * `worker/awards`. They are two independent implementations of one spec; neither is the
 * reference and neither may be corrected by reading the other's source.
 */

/** The versioned namespace shared by every stream. Bumping "v1" invalidates every ceremony. */
const LABEL_PREFIX = 'inclusivcup/v1';

/**
 * The single pity stream. Story 6.3 defines the LABEL only — the pity ALGORITHM (**FR-28**'s pity
 * roulette) is Story 6.7's, in {@link ./pity}.
 *
 * ⚠ THIS SIDE NAMED NO FR AT ALL UNTIL STORY 6.7, while Go's mirror named the WRONG one (`FR-26`,
 * which is anti-sweep plus the luck meter). Both now say FR-28, so the two halves of the seam make
 * the same claim about the same constant — the asymmetry was measured at 6.7's contexting, along
 * with the fact that `FR-28` appeared ZERO times anywhere in the seam before this comment.
 *
 * ⛔ THE VALUE IS FROZEN. It is the HMAC message prefix for every consolation draw ever published;
 * moving one byte of it keys a different stream and invalidates every ceremony. Only the trace
 * above changed.
 */
export const PITY_LABEL = `${LABEL_PREFIX}/pity`;

/**
 * The domain separator for Stage-1 spin S.
 *
 * `spin` is 1-BASED and formatted as plain decimal with NO padding: 1 -> ".../spin/1",
 * 12 -> ".../spin/12". `String(spin)` — never `toLocaleString()`, which inserts a thousands
 * separator in most locales (`"1,000"`) and would key a different stream on a different machine.
 *
 * A non-integer or non-positive spin is a programmer error, not a business refusal: it throws
 * rather than producing a label that would silently key a valid-looking but wrong stream.
 *
 * ⭐ The UPPER bound is not decoration. `Number.isInteger(1e21)` is `true`, and `String(1e21)` is
 * `"1e+21"` — so without it this returns `".../spin/1e+21"`, scientific notation, which is exactly
 * the formatting divergence the paragraph above spends five lines guarding against. Go's
 * `strconv.Itoa` on an `int` never does that. Measured at the 6.3 code review.
 */
export function stage1Label(spin: number): string {
  if (!Number.isInteger(spin) || spin < 1 || spin > Number.MAX_SAFE_INTEGER) {
    throw new RangeError(
      `roulette: stage1 spin must be an integer in [1, ${Number.MAX_SAFE_INTEGER}] (1-based), got ${String(spin)}`,
    );
  }
  return `${LABEL_PREFIX}/stage1/spin/${String(spin)}`;
}
